samples/switchless/README.md

```diff
@@ -0,0 +1,184 @@
+
+With the current implementation, we recommend users to avoid using more host worker threads than the minimum of:
+
+1. the number of simultaneously active enclave threads, and
+2. the number of cores that are potentially available to host worker threads.
+
+For example, on a 4-core machine, if the number of the simultaneously active enclave threads is 2, and there is no
+host threads other than the two threads making ECALLs and the switchless worker threads, both 1) and 2) would be 2. So we recommend setting the number of host worker threads to 2.
+
+The exception to the above rule happens when 1) or 2) is zero or negative. For example, if the host starts two more
+additional threads that are expected to be active along with the two enclave threads, the number of cores available
+to the worker threads is actually 0, and the minimum of 1) and 2) would be 0. In this case, we recommend setting
+the number of host worker threads to 1 nevertheless, to ensure switchless calls are serviced by at least one thread.
+
+The above recommendation may change when we modify the behavior of worker threads in the future.
+
+## About the EDL
+
+In this sample, we pretend the enclave doesn't know addition. It relies on a host function `host_increment` to
+increment a number by 1, and repeat calling it `N` times to add `N` to a given number. Since `host_increment` is
+short and called frequently, it is appropriate to make it a switchless function.
+
+First we need to define the functions we want to call between the host and the enclave. To do this we create a `switchless.edl` file:
+
+```edl
+enclave {
+    trusted {
+        public void enclave_add_N([in, out] int* m, int n);
+    };
+
+    untrusted {
+        void host_increment([in, out] int* m) transition_using_threads;
+    };
+};
+```
+
+Function `host_increment`'s declaration ends with keyword `transition_using_threads`, indicating it should be called switchlessly at run time. However, this a best-effort directive. OpenEnclave runtime may still choose to fall back to a tradition OCALL if switchless call resources are unavailable, e.g., the enclave is not configured as switchless-capable, or the host worker threads are busy servicing other switchless OCALLs.
+
+To generate the functions with the marshaling code, the `oeedger8r` tool is called in both the host and enclave directories from their Makefiles. For example:
+
+```bash
+cd host
+oeedger8r ../switchless.edl --untrusted --experimental
+```
+
+`oeedger8r` needs the command line flag `--experimental` to be able to recognize the keyword `transition_using_threads`.
+
+## About the host
+
+The host first defines a structure specifically for configuring switchless calls. In this case, we specify the
+first field `1` as the number of host worker threads for switchless OCALLs. In this example, 1) There is at most
+1 enclave thread all the time, and 2) The number of cores available to the host worker threads is  1 (assuming a
```

Comment:
I would change "2)" to something like "The number of cores available to the host worker threads is unknown, and so we use 1 as explained above."

Rationale: We should not have sample code that "assumes" a 2-core or a 4-core machine.

Reply:
Fixed